### ISSUE TITLE
iPython cluster resource usage

### ISSUE THREAD

**Issue by matthdsm** [chapmanb/bcbio-nextgen#1542]
Hi Brad,

We're seeing a rather complicated problem. We notice that when we're trying to use our cluster to full capacity, the iPyhton jobs keep throwing timeout errors.

To illustrate:

Our cluster consists of 2 nodes, with respectively 8 and 16 cores, and 2GB ram per core. When we start a bcbio run on 16 cores (memory constrained) , the jobs get splitted nicely over the two nodes, and everything runs smootly and fast. When the first parallel job (mapping etc.)  is finished and iPython tries to submit the following jobs (variant calling etc.) the analysis shuts down, noting the new iPython jobs have timed out. I think this is because the previous jobs take a while to exit fully, while still holding the allotted cores and memory occupied. This causes one of the new jobs to remain queued for some time, and resulting in a timeout. In practice, this means we have to restart the analysis 2 times for it to finish, if we want to use our cluster to it's full potential.

I we start the analysis on just 8 cores ( so more resources remain available), the cascade of jobs runs perfectly.

I suppose a fix for this issue would be to speed up the exit of previous jobs, or to increase the timeout on new jobs. Or if possible in any way, perhaps wait to submit any new jobs untill the others have finished.

sorry for the wall of text, but I'd like to hear your thoughts on this.

Thanks!
M

**Comment by roryk** [comment]
Hi @matthdsm,

Does bumping up the ``--timeout`` parameter help? By default it waits for a short amount of time for resources to be available, setting that to something higher in minutes might fix the problem. 

On some schedulers we have trouble shutting down finished jobs at the end, which could be contributing to the problem. Which scheduler are you using?

For the job that is timing out, is it requesting too many resources for the cluster? We estimate what the cluster looks like by submitting one job and looking at what the machine has for resources and then assuming all of the machines have about those resources available. In a heterogeneous cluster like this, we might be assuming too many resources exist and ask for too much. If you look at the job submission script for the step that is timing out, is it asking for more resources than your cluster has?

Hope any of that helps.

**Comment by roryk** [comment]
You can see what bcbio estimated was available by looking in `provenance/system-ipython-general.yaml` . If it says 16 cores in there it might be over-requesting resources. If you switch that to 8 it should run. But at that point I think you're better off just running the single job locally on the 16 core node since you'll use the same amount of cores.

**Comment by matthdsm** [comment]
Hi @roryk, 

Thanks for your 2 cents! We'll try to increase the timeout, and perhaps use the retry option. Guess I missed those in the manual. I'll get back to you when we've got some results!

M

**Comment by matthdsm** [comment]
Hi @roryk 

As promised, the results of our latest run. But first, some more info. We're running a torque cluster. I suppose this is one of the clusters where you have issues closing finished jobs? I think this is exactly the problem we're having. Strange thing is, it's a bit hit and miss, sometimes everything runs smoothly, other times, thing go to hell right after the first ipython jobs have finished.

Could we be missing something in the configuration of our cluster? Or could there be some port used by iPython that is blocked by the firewall between our nodes?

tl;dr: Adding the `--retries` and `--timeout` option didn't help. The jobs timed out trying to start the variant call postprocessing (QC,gemini, etc.)

M

**Comment by roryk** [comment]
Hi Matt,

If the cluster is just two nodes, one with 16 cores and one with 8 cores, you're probably better off just running bcbio in local mode on the 16 core node. bcbio will look at either the 8 core or the 16 core machine to figure out what resources are available and either overschedule (if jt thinks both have 16 cores) or underschedule (if it thinks both of 8 cores). Either way, you'll be operating for the most part using a maximum of 16 cores so you won't be losing much.

**Comment by matthdsm** [comment]
Hi @roryk 


We're running the bcbio jobs as "local" now. It's too bad we can't make use of the iPython possibilities, but at least we're able to finish a run without any issues. thanks for the help.

M
